Numpy 1.26.0 doesn't work on freebsd

### Steps to reproduce:

```
python3.11 -m venv npv
. npv/bin/activate
pip install numpy==1.26.0
python
>>> import numpy
```

### Error message:

```shell
Python 3.11.5 (main, Sep  7 2023, 03:42:48) [Clang 14.0.5 (https://github.com/llvm/llvm-project.git llvmorg-14.0.5-0-gc12386 on freebsd13
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "/usr/home/lmoskala/npv/lib/python3.11/site-packages/numpy/__init__.py", line 130, in <module>
    from numpy.__config__ import show as show_config
  File "/usr/home/lmoskala/npv/lib/python3.11/site-packages/numpy/__config__.py", line 4, in <module>
    from numpy.core._multiarray_umath import (
  File "/usr/home/lmoskala/npv/lib/python3.11/site-packages/numpy/core/__init__.py", line 100, in <module>
    from . import _add_newdocs
  File "/usr/home/lmoskala/npv/lib/python3.11/site-packages/numpy/core/_add_newdocs.py", line 4973, in <module>
    add_newdoc('numpy.core._multiarray_tests', 'format_float_OSprintf_g',
  File "/usr/home/lmoskala/npv/lib/python3.11/site-packages/numpy/core/function_base.py", line 543, in add_newdoc
    new = getattr(__import__(place, globals(), {}, [obj]), obj)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ImportError: /usr/home/lmoskala/npv/lib/python3.11/site-packages/numpy/core/_multiarray_tests.cpython-311.so: Undefined symbol "ccosl"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/home/lmoskala/npv/lib/python3.11/site-packages/numpy/__init__.py", line 135, in <module>
    raise ImportError(msg) from e
ImportError: Error importing numpy: you should not try to import numpy from
        its source directory; please exit the numpy source tree, and relaunch
        your python interpreter from there.
>>>
```


### Additional information:

FreeBSD 13.2 with python3.11.5 (python 3.11.2 has the same issue).

If I do `pip install numpy==1.24.1`, then `import numpy` succeeds. Numpy 1.25.2 also works. So I'd guess that somewhere between 1.25.2 and 1.26.0 #22850 stopped working? of a new function appeared?

I'll try to provide any informations needed.

Thanks @LukaszMoskala. So the check passes. Something is very badly wrong with those checks, if I look in `build/meson-logs/meson-log.txt` on Linux I see:

```
Running compile:
Working directory:  /home/rgommers/code/numpy/build/meson-private/tmpgkbtw7os
Code:
 
        #define ccosl meson_disable_define_of_ccosl
        
        #include <limits.h>
        #undef ccosl
        
        #ifdef __cplusplus
        extern "C"
        #endif
        char ccosl (void);
        
        #if defined __stub_ccosl || defined __stub___ccosl
        fail fail fail this function is not going to work
        #endif
        
        int main(void) {
          return ccosl ();
        }
-----------
Command line: `/home/rgommers/mambaforge/envs/numpy-dev/bin/x86_64-conda-linux-gnu-cc -L/home/rgommers/mambaforge/envs/numpy-dev/lib /home/rgommers/code/numpy/build/meson-private/tmpgkbtw7os/testfile.c -o /home/rgommers/code/numpy/build/meson-private/tmpgkbtw7os/output.exe -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/rgommers/mambaforge/envs/numpy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/rgommers/mambaforge/envs/numpy-dev/include -D_FILE_OFFSET_BITS=64 -O0 -std=c99 -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/rgommers/mambaforge/envs/numpy-dev/lib -Wl,-rpath-link,/home/rgommers/mambaforge/envs/numpy-dev/lib` -> 1
stderr:
/home/rgommers/code/numpy/build/meson-private/tmpgkbtw7os/testfile.c:10:14: warning: conflicting types for built-in function 'ccosl'; expected '_Complex long double(_Complex long double)' [-Wbuiltin-declaration-mismatch]
   10 |         char ccosl (void);
      |              ^~~~~
/home/rgommers/code/numpy/build/meson-private/tmpgkbtw7os/testfile.c:5:1: note: 'ccosl' is declared in header '<complex.h>'
    4 |         #include <limits.h>
  +++ |+#include <complex.h>
    5 |         #undef ccosl
/home/rgommers/mambaforge/envs/numpy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/11.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: /tmp/ccQflKnA.o: in function `main':
testfile.c:(.text.main+0x6): undefined reference to `ccosl'
collect2: error: ld returned 1 exit status
```
but the check passes!

Once I improve the check by adding the missing `complex.h` include:
```diff
--- a/numpy/core/meson.build
+++ b/numpy/core/meson.build
@@ -181,7 +181,7 @@ foreach func: c99_complex_funcs
   if cc.has_function(func_single)
     cdata.set10('HAVE_' + func_single.to_upper(), true)
   endif
-  if cc.has_function(func_longdouble)
+  if cc.has_function(func_longdouble, prefix: '#include <complex.h>')
     cdata.set10('HAVE_' + func_longdouble.to_upper(), true)
   endif
 endforeach
```
it starts failing with:

```
Running compile:
Working directory:  /home/rgommers/code/numpy/build/meson-private/tmppjsv4dqy
Code:
 #include <complex.h>
#include <limits.h>

        #if defined __stub_ccosl || defined __stub___ccosl
        fail fail fail this function is not going to work
        #endif
        
int main(void) {
            void *a = (void*) &ccosl;
            long long b = (long long) a;
            return (int) b;
        }
-----------
Command line: `/home/rgommers/mambaforge/envs/numpy-dev/bin/x86_64-conda-linux-gnu-cc -L/home/rgommers/mambaforge/envs/numpy-dev/lib /home/rgommers/code/numpy/build/meson-private/tmppjsv4dqy/testfile.c -o /home/rgommers/code/numpy/build/meson-private/tmppjsv4dqy/output.exe -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/rgommers/mambaforge/envs/numpy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/rgommers/mambaforge/envs/numpy-dev/include -D_FILE_OFFSET_BITS=64 -O0 -std=c99 -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/rgommers/mambaforge/envs/numpy-dev/lib -Wl,-rpath-link,/home/rgommers/mambaforge/envs/numpy-dev/lib` -> 1
stderr:
/home/rgommers/mambaforge/envs/numpy-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/11.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: /tmp/ccAXOZ80.o: in function `main':
testfile.c:(.text.main+0x7): undefined reference to `ccosl'
collect2: error: ld returned 1 exit status
```

Okay, way more is broken here with the `cc.has_function` checks, it just happens to work in our CI because we don't test anything on a non-C99-complete platform. This needs fixing asap.

--- a/numpy/core/meson.build
+++ b/numpy/core/meson.build
@@ -148,20 +148,32 @@ endforeach
 # Mandatory functions: if not found, fail the build
 # Some of these can still be blocklisted if the C99 implementation
 # is buggy, see numpy/core/src/common/npy_config.h
-mandatory_funcs = [
+mandatory_math_funcs = [
   'sin', 'cos', 'tan', 'sinh', 'cosh', 'tanh', 'fabs',
   'floor', 'ceil', 'sqrt', 'log10', 'log', 'exp', 'asin',
   'acos', 'atan', 'fmod', 'modf', 'frexp', 'ldexp',
   'expm1', 'log1p', 'acosh', 'asinh', 'atanh',
-  'rint', 'trunc', 'exp2',
-  'copysign', 'nextafter', 'strtoll', 'strtoull', 'cbrt',
+  'rint', 'trunc', 'exp2', 'copysign', 'nextafter', 'cbrt',
   'log2', 'pow', 'hypot', 'atan2',
-  'csin', 'csinh', 'ccos', 'ccosh', 'ctan', 'ctanh',
-  'creal', 'cimag', 'conj'
 ]
-foreach func: mandatory_funcs
-  if not cc.has_function(func)
-    error('Function `{func}` not found')
+foreach func: mandatory_math_funcs
+  if not cc.has_function(func, prefix: '#include <math.h>', dependencies: m_dep)
+    error(f'Function `@func@` not found')
+  endif
+endforeach
+
+mandatory_complex_math_funcs = [
+  'csin', 'csinh', 'ccos', 'ccosh', 'ctan', 'ctanh', 'creal', 'cimag', 'conj'
+]
+foreach func: mandatory_complex_math_funcs
+  if not cc.has_function(func, prefix: '#include <complex.h>', dependencies: m_dep)
+    error(f'Function `@func@` not found')
+  endif
+endforeach
+
+foreach func: ['strtoll', 'strtoull']
+  if not cc.has_function(func, prefix: '#include <stdlib.h>')
+    error(f'Function `@func@` not found')
   endif
 endforeach
 
@@ -175,13 +187,13 @@ c99_complex_funcs = [
 foreach func: c99_complex_funcs
   func_single = func + 'f'
   func_longdouble = func + 'l'
-  if cc.has_function(func)
+  if cc.has_function(func, prefix: '#include <complex.h>', dependencies: m_dep)
     cdata.set10('HAVE_' + func.to_upper(), true)
   endif
-  if cc.has_function(func_single)
+  if cc.has_function(func_single, prefix: '#include <complex.h>', dependencies: m_dep)
     cdata.set10('HAVE_' + func_single.to_upper(), true)
   endif
-  if cc.has_function(func_longdouble)
+  if cc.has_function(func_longdouble, prefix: '#include <complex.h>', dependencies: m_dep)
     cdata.set10('HAVE_' + func_longdouble.to_upper(), true)
   endif
 endforeach
@@ -190,7 +202,7 @@ endforeach
 # libnpymath as a C99 compat layer, these may still be relevant.
 c99_macros = ['isfinite', 'isinf', 'isnan', 'signbit']
 foreach macro: c99_macros
-  if cc.has_function(macro)
+  if cc.has_function(macro, prefix: '#include <math.h>', dependencies: m_dep)
     cdata.set10('NPY_HAVE_DECL_' + macro.to_upper(), true)
     if not cc.has_header_symbol('Python.h', macro, dependencies: py_dep)
       # Add in config.h as well, except if it will clash with Python's config.h content

--- a/numpy/core/src/npymath/npy_math_complex.c.src
+++ b/numpy/core/src/npymath/npy_math_complex.c.src
@@ -163,7 +163,9 @@ npy_carg@c@(@ctype@ z)
 #define SCALED_CEXP_LOWERL 11357.216553474703895L
 #define SCALED_CEXP_UPPERL 22756.021937783004509L
 
-#if !defined(HAVE_CEXP@C@)
+#if !defined(HAVE_CSINH@C@) || \
+    !defined(HAVE_CCOSH@C@) || \
+    !defined(HAVE_CEXP@C@)
 
 static
 @ctype@

--- a/numpy/core/tests/test_multiarray.py
+++ b/numpy/core/tests/test_multiarray.py
@@ -9698,9 +9698,12 @@ def test_ragged_comparison_fails(op):
 def test_npymath_complex(fun, npfun, x, y, test_dtype):
     # Smoketest npymath functions
     z = test_dtype(complex(x, y))
-    got = fun(z)
-    expected = npfun(z)
-    assert_allclose(got, expected)
+    with np.errstate(invalid='ignore'):
+        # Fallback implementations may emit a warning for +-inf (see gh-24876):
+        #     RuntimeWarning: invalid value encountered in absolute
+        got = fun(z)
+        expected = npfun(z)
+        assert_allclose(got, expected)
 
 
 def test_npymath_real():

--- a/numpy/core/tests/test_numeric.py
+++ b/numpy/core/tests/test_numeric.py
@@ -916,6 +916,11 @@ class TestFloatExceptions:
     @pytest.mark.skipif(IS_WASM, reason="no wasm fp exception support")
     @pytest.mark.parametrize("typecode", np.typecodes["AllFloat"])
     def test_floating_exceptions(self, typecode):
+        if 'bsd' in sys.platform and typecode in 'gG':
+            pytest.skip(reason="Fallback impl for (c)longdouble may not raise "
+                               "FPE errors as expected on BSD OSes, "
+                               "see gh-24876, gh-23379")
+
         # Test basic arithmetic function errors
         with np.errstate(all='raise'):
             ftype = obj2sctype(typecode)

--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -1645,6 +1645,8 @@ class TestSpecialFloats:
                           np.array(1200.0, dtype='d'))
 
     @pytest.mark.skipif(IS_WASM, reason="fp errors don't work in wasm")
+    @pytest.mark.skipif('bsd' in sys.platform,
+            reason="fallback implementation may not raise, see gh-2487")
     def test_cosh(self):
         in_ = [np.nan, -np.nan, np.inf, -np.inf]
         out = [np.nan, np.nan, np.inf, np.inf]

--- a/numpy/core/tests/test_umath_complex.py
+++ b/numpy/core/tests/test_umath_complex.py
@@ -576,8 +576,8 @@ class TestSpecialComplexAVX:
                             complex(0.,     0.),
                             complex(np.nan, np.nan),
                             complex(np.nan, np.nan)], dtype=astype)
-        assert_equal(np.abs(arr[::stride]), abs_true[::stride])
         with np.errstate(invalid='ignore'):
+            assert_equal(np.abs(arr[::stride]), abs_true[::stride])
             assert_equal(np.square(arr[::stride]), sq_true[::stride])
 
 class TestComplexAbsoluteAVX:

--- a/tools/ci/cirrus_arm.yml
+++ b/tools/ci/cirrus_arm.yml
@@ -96,7 +96,7 @@ macos_arm64_test_task:
 
     RUNNER_OS="macOS"
     SDKROOT=/Applications/Xcode-14.0.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk
-    
+
     # NOTE: OpenBLAS is not used in this job; if that's done in the future, ensure
     # PKG_CONFIG_PATH points to the directory containing the openblas.pc file
     # that's installed with the cibw_before_build.sh command.
@@ -110,8 +110,50 @@ macos_arm64_test_task:
 
     pip install -r build_requirements.txt
     pip install pytest pytest-xdist hypothesis typing_extensions
-    
+
     spin build -- -Dallow-noblas=true
     spin test -j auto
 
+    ccache -s
+
+
+freebsd_test_task:
+  use_compute_credits: $CIRRUS_USER_COLLABORATOR == 'true'
+  compute_engine_instance:
+    image_project: freebsd-org-cloud-dev
+    image: family/freebsd-13-2
+    platform: freebsd
+    cpu: 1
+    memory: 4G
+
+  install_devtools_script: |
+    pkg install -y git bash ninja ccache
+
+  <<: *MODIFIED_CLONE
+
+  ccache_cache:
+    folder: .ccache
+    populate_script:
+      - mkdir -p .ccache
+    fingerprint_key: ccache-freebsd
+
+  prepare_env_script: |
+    # Create a venv (the `source` command needs bash, not the default sh shell)
+    chsh -s /usr/local/bin/bash
+    python -m venv .venv
+    source .venv/bin/activate
+    # Minimal build and test requirements
+    python -m pip install -U pip
+    python -m pip install meson-python Cython pytest hypothesis
+
+  build_script: |
+    chsh -s /usr/local/bin/bash
+    source .venv/bin/activate
+    python -m pip install . --no-build-isolation -v -Csetup-args="-Dallow-noblas=true"
+
+  test_script: |
+    chsh -s /usr/local/bin/bash
+    source .venv/bin/activate
+    cd tools
+    python -m pytest --pyargs numpy -m "not slow"
     ccache -s

